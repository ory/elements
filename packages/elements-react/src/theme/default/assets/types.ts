// Copyright Â© 2025 Ory Corp
// SPDX-License-Identifier: Apache-2.0

export type SVGIcon = React.FunctionComponent<
  React.ComponentProps<"svg"> & { size?: number }
>
