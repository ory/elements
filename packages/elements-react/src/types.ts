// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import {
  UiNode,
  UiNodeAnchorAttributes,
  UiNodeImageAttributes,
  UiNodeInputAttributes,
  UiNodeTextAttributes,
} from "@ory/client-fetch"
import {
  ComponentPropsWithoutRef,
  FormEventHandler,
  MouseEventHandler,
  PropsWithChildren,
} from "react"
import { ButtonVariants } from "./theme/default/components/form/button"

export type OryNodeButtonProps = {
  attributes: UiNodeInputAttributes
  node: UiNode
} & Omit<ComponentPropsWithoutRef<"button">, "children"> &
  ButtonVariants

export type OryCurrentIdentifierProps = {
  attributes: UiNodeInputAttributes
  node: UiNode
  onClick?: () => void
  href?: string
} & Omit<ComponentPropsWithoutRef<"button">, "children" | "onClick">

export type OryNodeAnchorProps = {
  attributes: UiNodeAnchorAttributes
  node: UiNode
} & Omit<ComponentPropsWithoutRef<"a">, "children">

export type OryNodeLabelProps = {
  attributes: UiNodeInputAttributes
  node: UiNode
} & ComponentPropsWithoutRef<"label">

export type OryNodeTextProps = {
  attributes: UiNodeTextAttributes
  node: UiNode
}

export type OryCardLogoProps = Record<string, never>

/**
 * Props for the AuthMethodListItem component. This component is used
 * to render a single auth method in the AuthMethodList component.
 */
export type OryCardAuthMethodListItemProps = {
  onClick: () => void
  group: string
}

export type OryNodeImageProps = {
  attributes: UiNodeImageAttributes
  node: UiNode
}

/**
 * A generic type for the form values.
 */
export type FormValues = Record<string, string | boolean | number | undefined>

export type OryFormRootProps = ComponentPropsWithoutRef<"form"> & {
  onSubmit: FormEventHandler<HTMLFormElement>
}

export type OryNodeInputProps = {
  attributes: UiNodeInputAttributes
  node: UiNode
  onClick?: MouseEventHandler
}

export type OryFormSectionContentProps = PropsWithChildren<{
  title?: string
  description?: string
}>

export type OryFormSectionFooterProps = PropsWithChildren
