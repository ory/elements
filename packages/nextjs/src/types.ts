// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

export type QueryParams = { [key: string]: string | string[] | undefined }

export const initOverrides: RequestInit = {
  cache: "no-cache",
}

export type FlowParams = {
  id: string
  cookie: string | undefined
  return_to: string
}
