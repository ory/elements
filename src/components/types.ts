import {ReactElement} from "react";

export type Children = string | string[] | ReactElement | ReactElement[];
