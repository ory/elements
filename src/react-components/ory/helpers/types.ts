// Copyright Â© 2022 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import {
  LoginFlow,
  RecoveryFlow,
  RegistrationFlow,
  SettingsFlow,
  VerificationFlow,
} from "@ory/client"

export type SelfServiceFlow =
  | LoginFlow
  | RecoveryFlow
  | RegistrationFlow
  | SettingsFlow
  | VerificationFlow
