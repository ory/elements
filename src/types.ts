import {
  SelfServiceLoginFlow,
  SelfServiceRecoveryFlow,
  SelfServiceRegistrationFlow,
  SelfServiceSettingsFlow,
  SelfServiceVerificationFlow,
} from "@ory/client"

export type SelfServiceFlow =
  | SelfServiceLoginFlow
  | SelfServiceRecoveryFlow
  | SelfServiceRegistrationFlow
  | SelfServiceSettingsFlow
  | SelfServiceVerificationFlow
