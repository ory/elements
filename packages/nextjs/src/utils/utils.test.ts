// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

// utils.test.ts

import {
  onValidationError,
  toFlowParams,
  processSetCookieHeaders,
  filterRequestHeaders,
  joinUrlPaths,
} from "./utils"
import { OryConfig, QueryParams } from "../types"

describe("onValidationError", () => {
  it("should return the same value passed to it", () => {
    const value = { key: "value" }
    expect(onValidationError(value)).toBe(value)
  })
})

describe("toFlowParams", () => {
  it("should return FlowParams with id, cookie, and return_to", async () => {
    const params: QueryParams = {
      ["flow"]: "some-flow-id",
      ["return_to"]: "https://example.com/return",
    }
    const getCookieHeader = jest.fn().mockResolvedValue("some-cookie-value")

    const result = await toFlowParams(params, getCookieHeader)

    expect(result).toEqual({
      id: "some-flow-id",
      cookie: "some-cookie-value",
      return_to: "https://example.com/return",
    })
    expect(getCookieHeader).toHaveBeenCalled()
  })
})

describe("processSetCookieHeaders", () => {
  ;[
    {
      name: "should respect forwarded headers",
      protocol: "http",
      forwardedProtocol: "https",
      host: "console.ory.sh",
      forwardedHost: "api.console.ory.sh",
      headers: new Headers([
        ["Set-Cookie", "sessionid=abc123; Path=/; HttpOnly"],
      ]),
    },
    {
      name: "should respect regular headers",
      protocol: "https",
      host: "console.ory.sh",
      headers: new Headers([
        ["set-cookie", "sessionid=abc123; Path=/; HttpOnly"],
      ]),
    },
    {
      name: "supports insecure",
      protocol: "http",
      host: "console.ory.sh",
      headers: new Headers([
        ["set-cookie", "sessionid=abc123; Path=/; HttpOnly"],
      ]),
    },
    {
      name: "supports multiple cookies comma separated",
      protocol: "http",
      host: "console.ory.sh",
      headers: new Headers([
        [
          "set-cookie",
          "sessionid1=abc123; Path=/; HttpOnly, sessionid2=123abc; Path=/abc; HttpOnly",
        ],
      ]),
    },
    {
      name: "supports multiple cookies in record",
      protocol: "http",
      host: "console.ory.sh",
      headers: new Headers([
        ["set-cookie", "sessionid1=abc123; Path=/; HttpOnly"],
        ["set-cookiE", "sessionid2=123abc; Path=/abc; HttpOnly"],
      ]),
    },
  ].forEach(
    ({
      name,
      protocol,
      forwardedProtocol,
      host,
      forwardedHost,
      headers,
      forceCookieDomain,
    }: {
      name: string
      protocol: string
      forwardedProtocol?: string
      host: string
      forwardedHost?: string
      headers: Headers
      forceCookieDomain?: string
    }) => {
      test(name, () => {
        const options: OryConfig = {
          forceCookieDomain,
        }
        const requestHeaders = new Headers()
        requestHeaders.set("host", host)
        if (forwardedProtocol) {
          requestHeaders.set("x-forwarded-proto", forwardedProtocol)
        }
        if (forwardedHost) {
          requestHeaders.set("x-forwarded-host", forwardedHost)
        }

        const fetchResponse = new Response(null, {
          headers,
        })

        const result = processSetCookieHeaders(
          protocol,
          fetchResponse,
          options,
          requestHeaders,
        )

        expect(result).toMatchSnapshot()
      })
    },
  )
})

describe("filterRequestHeaders", () => {
  it("should filter headers based on default and additional headers", () => {
    const headers = new Headers()
    headers.set("authorization", "Bearer token")
    headers.set("content-type", "application/json")
    headers.set("cookie", "sessionid=abc123")
    headers.set("x-custom-header", "custom-value")
    headers.set("x-ignore-header", "custom-value")

    const forwardAdditionalHeaders = ["x-custom-header"]

    const result = filterRequestHeaders(headers, forwardAdditionalHeaders)

    expect(result.get("authorization")).toBe("Bearer token")
    expect(result.get("content-type")).toBe("application/json")
    expect(result.get("x-custom-header")).toBe("custom-value")
    expect(result.has("x-ignore-header")).toBe(false)
  })

  it("should filter headers based on default headers only when additional headers are not provided", () => {
    const headers = new Headers()
    headers.set("authorization", "Bearer token")
    headers.set("content-type", "application/json")
    headers.set("cookie", "sessionid=abc123")
    headers.set("x-custom-header", "custom-value")

    const result = filterRequestHeaders(headers)

    expect(result.get("authorization")).toBe("Bearer token")
    expect(result.get("content-type")).toBe("application/json")
    expect(result.has("x-custom-header")).toBe(false)
  })
})

describe("joinUrlPaths", () => {
  it("should correctly join base URL and relative URL", () => {
    const baseUrl = "https://example.com/api"
    const relativeUrl = "/v1/resource"

    const result = joinUrlPaths(baseUrl, relativeUrl)

    expect(result).toBe("https://example.com/api/v1/resource")
  })

  it("should handle base URL without trailing slash", () => {
    const baseUrl = "https://example.com/"
    const relativeUrl = "v1/resource"

    const result = joinUrlPaths(baseUrl, relativeUrl)

    expect(result).toBe("https://example.com/v1/resource")
  })

  it("should handle relative URL with full URL", () => {
    const baseUrl = "https://example.com/api"
    const relativeUrl = "https://another.com/resource"

    const result = joinUrlPaths(baseUrl, relativeUrl)

    expect(result).toBe("https://another.com/api/resource")
  })
})
