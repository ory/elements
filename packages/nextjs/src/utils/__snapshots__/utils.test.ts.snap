// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processSetCookieHeaders should respect forwarded headers 1`] = `
[
  "sessionid=abc123; Domain=ory.sh; Path=/; HttpOnly; Secure",
]
`;

exports[`processSetCookieHeaders should respect regular headers 1`] = `
[
  "sessionid=abc123; Domain=ory.sh; Path=/; HttpOnly",
]
`;

exports[`processSetCookieHeaders supports insecure 1`] = `
[
  "sessionid=abc123; Domain=ory.sh; Path=/; HttpOnly",
]
`;

exports[`processSetCookieHeaders supports multiple cookies comma separated 1`] = `
[
  "sessionid1=abc123; Domain=ory.sh; Path=/; HttpOnly",
  "sessionid2=123abc; Domain=ory.sh; Path=/abc; HttpOnly",
]
`;

exports[`processSetCookieHeaders supports multiple cookies in record 1`] = `
[
  "sessionid1=abc123; Domain=ory.sh; Path=/; HttpOnly",
  "sessionid2=123abc; Domain=ory.sh; Path=/abc; HttpOnly",
]
`;
